actors/miner/src/lib.rs

```diff
@@ -3652,54 +3791,229 @@ where
+}
+
+fn update_replica_states<BS>(
+    rt: &impl Runtime,
+    updates_by_deadline: &BTreeMap<u64, Vec<ReplicaUpdateStateInputs>>,
+    expected_count: usize,
+    sectors: &mut Sectors<BS>,
+    sector_size: SectorSize,
+) -> Result<(PowerPair, TokenAmount), ActorError>
+where
+    BS: Blockstore,
+{
+    let rew = request_current_epoch_block_reward(rt)?;
+    let pow = request_current_total_power(rt)?;
+    let circulating_supply = rt.total_fil_circ_supply();
+    let pledge_inputs = NetworkPledgeInputs {
+        network_qap: pow.quality_adj_power_smoothed,
+        network_baseline: rew.this_epoch_baseline_power,
+        circulating_supply,
+        epoch_reward: rew.this_epoch_reward_smoothed,
+    };
+    let mut power_delta = PowerPair::zero();
+    let mut pledge_delta = TokenAmount::zero();
+
+    rt.transaction(|state: &mut State, rt| {
+        let mut deadlines = state.load_deadlines(rt.store())?;
+        let mut new_sectors = Vec::with_capacity(expected_count);
+        // Process updates grouped by deadline.
+        for (&dl_idx, updates) in updates_by_deadline {
+            let mut deadline = deadlines.load_deadline(rt.store(), dl_idx)?;
+
+            let mut partitions = deadline
+                .partitions_amt(rt.store())
+                .with_context_code(ExitCode::USR_ILLEGAL_STATE, || {
+                    format!("failed to load partitions for deadline {}", dl_idx)
+                })?;
+
+            let quant = state.quant_spec_for_deadline(rt.policy(), dl_idx);
+
+            for update in updates {
+                // Compute updated sector info.
+                let new_sector_info = update_existing_sector_info(
+                    update.sector_info,
+                    &update.activated_data,
+                    &pledge_inputs,
+                    sector_size,
+                    rt.curr_epoch(),
+                );
+
+                let mut partition = partitions
+                    .get(update.partition)
+                    .with_context_code(ExitCode::USR_ILLEGAL_STATE, || {
+                        format!(
+                            "failed to load deadline {} partition {}",
+                            update.deadline, update.partition
+                        )
+                    })?
+                    .cloned()
+                    .ok_or_else(|| {
+                        actor_error!(
+                            not_found,
+                            "no such deadline {} partition {}",
+                            dl_idx,
+                            update.partition
+                        )
+                    })?;
+
+                // Note: replacing sectors one at a time in each partition is inefficient.
+                let (partition_power_delta, partition_pledge_delta) = partition
+                    .replace_sectors(
+                        rt.store(),
+                        &[update.sector_info.clone()],
+                        &[new_sector_info.clone()],
```

Comment:
Perf nit:

```suggestion
                        std::slice::from_ref(&update.sector_info),
                        std::slice::from_ref(&new_sector_info),
```

(you can transform a reference into a size-one slice with no copying)